### PR TITLE
8255493: Support for pre-generated java.lang.invoke classes in CDS dynamic archive

### PR DESCRIPTION
**Pull request by yminqi**
Hi, Please review

  When do dynamic dump, the pre-generated lambda form classes from java.lang.invoke are not stored in dynamic archive. The patch will regenerate the four holder classes, 
     "java.lang.invoke.Invokers$Holder",
     "java.lang.invoke.DirectMethodHandle$Holder",
     "java.lang.invoke.DelegatingMethodHandle$Holder",
     "java.lang.invoke.LambdaForm$Holder"
  which will include the versions in static archive and new loaded functions all together and stored in dynamic archive. New test case added.
  (Minor change to PrintSharedArchiveAtExit, which the index is not consecutive)

  Tests: tier1,tier2,tier3,tier4
  
  Thanks
  Yumin

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8255493](https://bugs.openjdk.java.net/browse/JDK-8255493): Support for pre-generated java.lang.invoke classes in CDS dynamic archive

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to 86247422479c7b315578f5b40afadf6aff082ef6
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to 61675bde18365a47d92f4ada02bf14e5aaee4b51

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3611/head:pull/3611` \
`$ git checkout pull/3611`

Update a local copy of the PR: \
`$ git checkout pull/3611` \
`$ git pull https://git.openjdk.java.net/jdk pull/3611/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3611`

View PR using the GUI difftool: \
`$ git pr show -t 3611`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3611.diff">https://git.openjdk.java.net/jdk/pull/3611.diff</a>

</details>

### CODE REVIEW

**Comment by yminqi** [comment]
> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [hotspot-dev](mailto:hotspot-dev@mail.openjdk.java.net):_
> 
> Hi Yumin,
> 
> On 22/04/2021 8:42 am, Yumin Qi wrote:
> 
> > Hi, Please review
> > When do dynamic dump, the pre-generated lambda form classes from java.lang.invoke are not stored in dynamic archive. The patch will regenerate the four holder classes,
> > "java.lang.invoke.Invokers$Holder",
> > "java.lang.invoke.DirectMethodHandle$Holder",
> > "java.lang.invoke.DelegatingMethodHandle$Holder",
> > "java.lang.invoke.LambdaForm$Holder"
> > which will include the versions in static archive and new loaded functions all together and stored in dynamic archive. New test case added.
> > (Minor change to PrintSharedArchiveAtExit, which the index is not consecutive)
> 
> Meta questions:
> - why are they not already stored?
> - why do we need them to be stored?
> - can we avoid hardwiring a set of classes that then tightly couples
> our code to the JDK code in a way that we may not easily detect breakages?
> 
> Thanks,
> David
Hi, David
1) When dump static archive, we stored the holder classes which were logged (bug https://bugs.openjdk.java.net/browse/JDK-8247536) during build time. The pre-generated lambda form classes were collected in jlink stage, then regenerated the holder classes which hold the LF functions, stored in static archive. That is, whatever LF forms generated during build are archive in static archive.
2) At runtime, user code Lambda will generate new classes which are not stored into dynamic archive. The holder classes are the four invoke class$holder classes, so we record the lambda format for dynamic dump, regenerate the holder classes which will host both static contents plus the newly loaded contents then stored in dynamic archive. There are two versions of those four holder classes, one is the static version, and the other (bigger) in dynamic, so if dynamic mapped, we use the dynamic versions which host more LF functions. 
3) We only regenerate the four holder classes and regardless of the others (species), those four holders host majority of the invoke functions. The names are hardcoded since there are no mapping from jdk to vm for this. Do we have a way to do this?

Thanks
Yumin

File: src/hotspot/share/memory/dynamicArchive.cpp [hunk 1]
```diff
@@ -367,6 +368,15 @@ void DynamicArchive::dump() {
     return;
   }
 
+  // regenerate lambdaform holder classes
+  if (LambdaFormInvokers::should_regenerate_holder_classes()) {
+    JavaThread* THREAD = JavaThread::current();
+    assert(THREAD->is_Java_thread(), "Should be JavaThread");
```
Comment (dholmes-ora) [inline src/hotspot/share/memory/dynamicArchive.cpp:374]: JavaThread::current() already asserts this.

File: src/hotspot/share/memory/dynamicArchive.cpp [hunk 2]
```diff
@@ -367,6 +368,15 @@ void DynamicArchive::dump() {
...
+    log_info(cds, dynamic)("Regenerate lambdaform holder classes ...");
+    LambdaFormInvokers::regenerate_holder_classes(THREAD);
```
Comment (dholmes-ora) [inline src/hotspot/share/memory/dynamicArchive.cpp:376]: What if this throws an exception?

File: src/hotspot/share/memory/dynamicArchive.cpp (hunk 1 above)
Comment (yminqi) [reply src/hotspot/share/memory/dynamicArchive.cpp:374]: Will remove the assert.
I will also move the check should_regenerate_holder_classes to LambdaFormInvokers::should_regenerate_holder_classes().

File: src/hotspot/share/memory/dynamicArchive.cpp (hunk 2 above)
Comment (yminqi) [reply src/hotspot/share/memory/dynamicArchive.cpp:376]: If there is exception in java side to regenerate the holder classes, there is a log for cds to indicate the regeneration failed, the exception is cleared and continue. It will not affect other parts.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp [hunk 3]
```diff
@@ -46,27 +47,56 @@
...
 
   Symbol* cds_name  = vmSymbols::jdk_internal_misc_CDS();
   Klass*  cds_klass = SystemDictionary::resolve_or_null(cds_name, THREAD);
   guarantee(cds_klass != NULL, "jdk/internal/misc/CDS must exist!");
 
+  log_info(cds)("Total lambdaform lines %d", _lambdaform_lines->length());
+  HandleMark hm(THREAD);
   int len = _lambdaform_lines->length();
   objArrayHandle list_lines = oopFactory::new_objArray_handle(vmClasses::String_klass(), len, CHECK);
+  if (list_lines == nullptr) {
```
Comment (iklam) [inline src/hotspot/share/cds/lambdaFormInvokers.cpp:90]: There's no need to check for null. If allocation fails, the CHECK in the previous line will force this function to return.
(same for the check on line 96).

File: src/hotspot/share/cds/lambdaFormInvokers.cpp [hunk 4]
```diff
@@ -97,9 +127,17 @@ void LambdaFormInvokers::regenerate_holder_classes(TRAPS) {
...
 
     char *class_name = java_lang_String::as_utf8_string(h_name());
+    if (class_name == nullptr) {
+      log_info(cds)("Out of memory when reloading classes, quit");
+      return;
+    }
     int len = h_bytes->length();
     // make a copy of class bytes so GC will not affect us.
     char *buf = resource_allocate_bytes(THREAD, len);
+    if (buf == nullptr) {
```
Comment (iklam) [inline src/hotspot/share/cds/lambdaFormInvokers.cpp:137]: `resource_allocate_bytes(THREAD, len)` by default uses `AllocFailStrategy::EXIT_OOM`. We should explicitly pass `RETURN_NULL` instead.

Also, the `resource_free_bytes` call below doesn't seem necessary. The buffer will be freed when the ResourceMark is popped.

File: src/hotspot/share/cds/metaspaceShared.cpp [hunk 5]
```diff
@@ -1450,6 +1453,19 @@ void MetaspaceShared::initialize_shared_spaces() {
...
+  if (DynamicDumpSharedSpaces) {
+    // Read stored LF format lines stored in static archive
+    Array<Array<char>*>* static_archive_invokers = LambdaFormInvokers::static_archive_invokers();
+    if (static_archive_invokers != nullptr) {
+      for (int i = 0; i < static_archive_invokers->length(); i++) {
+        Array<char>* line = static_archive_invokers->at(i);
+        char* str = line->adr_at(0);
+        LambdaFormInvokers::append_filtered(str);
+      }
+    }
```
Comment (iklam) [inline src/hotspot/share/cds/metaspaceShared.cpp:1466]: I think the above block should be moved into lambdaFormInvokers.cpp

**Review by iklam — CHANGES_REQUESTED** [review]

File: src/hotspot/share/cds/metaspaceShared.cpp (hunk 5 above)
Comment (yminqi) [reply src/hotspot/share/cds/metaspaceShared.cpp]: Done.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp (hunk 4 above)
Comment (yminqi) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:129]: Done.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp [hunk 6]
```diff
@@ -46,23 +47,44 @@
...
+    log_info(cds)("Nothing to regenerate for holder classes");
+    return;
+  }
   ResourceMark rm(THREAD);
 
   Symbol* cds_name  = vmSymbols::jdk_internal_misc_CDS();
   Klass*  cds_klass = SystemDictionary::resolve_or_null(cds_name, THREAD);
   guarantee(cds_klass != NULL, "jdk/internal/misc/CDS must exist!");
 
+  log_info(cds)("Total lambdaform lines %d", _lambdaform_lines->length());
```
[the remark this answers is not in the document]
Comment (yminqi) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:86]: Thanks for review! Will move into DEBUG_ONLY

File: src/hotspot/share/cds/lambdaFormInvokers.cpp [hunk 7]
```diff
@@ -97,15 +119,20 @@ void LambdaFormInvokers::regenerate_holder_classes(TRAPS) {
     assert(h_bytes != NULL, "Class bytes is NULL");
 
     char *class_name = java_lang_String::as_utf8_string(h_name());
+    if (class_name == nullptr) {
+      log_info(cds)("Out of memory when reloading classes, quit");
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (yminqi) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:125]: Will delete the result check.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp [hunk 8]
```diff
@@ -46,23 +47,44 @@
...
+  if (_lambdaform_lines == nullptr || _lambdaform_lines->length() == 0) {
+    log_info(cds)("Nothing to regenerate for holder classes");
+    return;
+  }
   ResourceMark rm(THREAD);
 
   Symbol* cds_name  = vmSymbols::jdk_internal_misc_CDS();
   Klass*  cds_klass = SystemDictionary::resolve_or_null(cds_name, THREAD);
   guarantee(cds_klass != NULL, "jdk/internal/misc/CDS must exist!");
+  DEBUG_ONLY(log_info(cds)("Total lambdaform lines %d", _lambdaform_lines->length());)
```
Comment (coleenp) [inline src/hotspot/share/cds/lambdaFormInvokers.cpp:85]: No sorry, I meant
log_debug(cds)("Total lambdaform lines %d", ...)
Since this is something you don't want with -Xlog:cds at the info level.  Same with line 77

File: src/hotspot/share/cds/archiveBuilder.cpp [hunk 9]
```diff
@@ -259,6 +260,7 @@ void ArchiveBuilder::gather_klasses_and_symbols() {
   log_info(cds)("    instance classes   = %5d", _num_instance_klasses);
   log_info(cds)("    obj array classes  = %5d", _num_obj_array_klasses);
   log_info(cds)("    type array classes = %5d", _num_type_array_klasses);
+  log_info(cds)("    symbols = %5d", _symbols->length());
```
Comment (calvinccheung) [inline src/hotspot/share/cds/archiveBuilder.cpp:263]: Please align the '=' at line 263 with the line above.

File: src/hotspot/share/cds/filemap.hpp [hunk 10]
```diff
@@ -237,6 +237,7 @@ class FileMapHeader: private CDSFileMapHeaderBase {
   bool   _use_full_module_graph;        // Can we use the full archived module graph?
   size_t _ptrmap_size_in_bits;          // Size of pointer relocation bitmap
   narrowOop _heap_obj_roots;            // An objArray that stores all the roots of archived heap objects
+
```
Comment (calvinccheung) [inline src/hotspot/share/cds/filemap.hpp:240]: I think you can revert the changes to this file since there are only couple of blank lines addition/deletion probably by mistake.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestDynamicRegenerateHolderClasses.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @test
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestDynamicRegenerateHolderClasses.java:26]: Please add a `@bug` line to this file.

**Review by calvinccheung — COMMENTED** [review]
Looks good. Just a few minor comments.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp (hunk 8 above)
Comment (yminqi) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:85]: Sorry, I misunderstood the comment.

File: src/hotspot/share/cds/archiveBuilder.cpp (hunk 9 above)
Comment (yminqi) [reply src/hotspot/share/cds/archiveBuilder.cpp:263]: Will reformat. Thanks.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestDynamicRegenerateHolderClasses.java (hunk 11 above)
Comment (yminqi) [reply test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestDynamicRegenerateHolderClasses.java:26]: Will update.

**Comment by yminqi** [comment]
Please hold, there is more clean up need to do: including file headers, empty lines etc.

File: src/hotspot/share/cds/dynamicArchive.cpp [hunk 12]
```diff
@@ -367,6 +368,12 @@ void DynamicArchive::dump() {
     return;
   }
 
+  // regenerate lambdaform holder classes
+  JavaThread* THREAD = JavaThread::current();
+  log_info(cds, dynamic)("Regenerate lambdaform holder classes ...");
+  LambdaFormInvokers::regenerate_holder_classes(THREAD);
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/dynamicArchive.cpp:374]: You still have not addressed the exception problem here. This call can return early with an exception pending due to line 89 in the called method, which uses CHECK. That exception will escape up to the Java code after the VM operation completes.

Note: If `regenerate_holder_classes` is changed to suppress all exceptions then it should not be declared TRAPS and there is no need to pass THREAD (manifest [Java]Thread::current() internally as needed).

File: src/hotspot/share/cds/lambdaFormInvokers.cpp [hunk 13]
```diff
@@ -147,5 +170,36 @@ void LambdaFormInvokers::reload_class(char* name, ClassFileStream& st, TRAPS) {
...
+
+void LambdaFormInvokers::dump_static_archive_invokers() {
+  if (_lambdaform_lines != nullptr && _lambdaform_lines->length() > 0) {
+    int count = _lambdaform_lines->length();
+    _static_archive_invokers = ArchiveBuilder::new_ro_array<Array<char>*>(count);
+    for (int i = 0; i < count; i++) {
+      char* str = _lambdaform_lines->at(i);
+      int len = (int)strlen(str);
+      Array<char>* line = ArchiveBuilder::new_ro_array<char>(len+1);
+      strcpy(line->adr_at(0), str); // copies terminating zero as well
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/lambdaFormInvokers.cpp:185]: Should use strncpy for safety (code checkers will flag this).

File: src/hotspot/share/classfile/systemDictionaryShared.cpp [hunk 14]
```diff
@@ -2261,7 +2274,7 @@ class SharedDictionaryPrinter : StackObj {
 
   void do_value(const RunTimeSharedClassInfo* record) {
     ResourceMark rm;
-    _st->print_cr("%4d: %s %s", (_index++), record->_klass->external_name(),
+    _st->print_cr("%4d: %s %s", ++_index, record->_klass->external_name(),
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/systemDictionaryShared.cpp:2277]: Not clear why the _index handling is changing.

File: src/hotspot/share/classfile/systemDictionaryShared.cpp [hunk 15]
```diff
@@ -2278,9 +2291,10 @@ class SharedLambdaDictionaryPrinter : StackObj {
       ResourceMark rm;
       Klass* k = record->proxy_klass_head();
       while (k != nullptr) {
-        _st->print_cr("%4d: %s %s", (++_index), k->external_name(),
+        _st->print_cr("%4d: %s %s", _index, k->external_name(),
                       class_loader_name_for_shared(k));
         k = k->next_link();
+        _index++;
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/systemDictionaryShared.cpp:2297]: Not clear why the _index handling is changing.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp [hunk 16]
```diff
@@ -46,23 +47,44 @@
...
   if (_lambdaform_lines == NULL) {
-    _lambdaform_lines = new GrowableArray<char*>(100);
+    _lambdaform_lines = new GrowableArrayCHeap<char*, mtClassShared>(150);
   }
   _lambdaform_lines->append(line);
 }
 
+void LambdaFormInvokers::append_filtered(char* line) {
+  for (int k = 0; k < NUM_FILTER; k++) {
+    if (strstr(line, filter[k]) != nullptr) {
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/lambdaFormInvokers.cpp:67]: Do you really want to match anywhere in the string? I'm assuming the line should start with the filter?

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Yumin,

A few comments below - the exception issue still remains.

I can't really comment on the actual archiving logic here as I'm not familiar enough with the design or code.

Thanks,
David

File: src/hotspot/share/cds/dynamicArchive.cpp (hunk 12 above)
Comment (yminqi) [reply src/hotspot/share/cds/dynamicArchive.cpp:374]: Thanks for pointing this out, I checked the function LambdaFormInvokers::regenerate_holder_classes again.  Looks the TRAPS can be removed. There are two cases: 
  1) If any allocation failed due to OOM, vm will exit except for the line 124:
           char *buf = resource_allocate_bytes(THREAD, len, AllocFailStrategy::RETURN_NULL);
           if (buf == nullptr) {
             log_info(cds)("Out of memory when reloading class %s, quit", class_name);
             return;
           }
2) CDS.generateLambdaFormHolderClasses may throw exception, the exception is supressed inside the function.
So I will remove the TRAPS.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp (hunk 13 above)
Comment (yminqi) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:185]: Will update with strncpy.

File: src/hotspot/share/classfile/systemDictionaryShared.cpp (hunk 14 above)
Comment (yminqi) [reply src/hotspot/share/classfile/systemDictionaryShared.cpp:2277]: This is found the printout of index is not consecutive between the two dictionaries. The previous index is printout, then increased then the next dictionary take the index and increased again. I will check again.

File: src/hotspot/share/cds/dynamicArchive.cpp (hunk 12 above)
Comment (dholmes-ora) [reply src/hotspot/share/cds/dynamicArchive.cpp:374]: Are you saying that this:

objArrayHandle list_lines = oopFactory::new_objArray_handle(vmClasses::String_klass(), len, CHECK);

won't throw an exception but will abort the VM? I'm not seeing that.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp (hunk 16 above)
Comment (yminqi) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:67]: You reminded me that only the invoker lines which contain four holder class be stored in static archive, not else. I will change that part. For the filtered lines, it starts with [LF_RESOLVE] .... so the comparison I used here to check if the line contains "java.lang.invoke.<Class$holder>".

File: src/hotspot/share/classfile/systemDictionaryShared.cpp (hunk 15 above)
Comment (yminqi) [reply src/hotspot/share/classfile/systemDictionaryShared.cpp]: I reverted this part. The problem is when construct the lambda dictionary library, I wrongly add one to the last dictionary's index.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp [hunk 17]
```diff
@@ -49,34 +49,42 @@
...
 
-void LambdaFormInvokers::regenerate_holder_classes(TRAPS) {
+void LambdaFormInvokers::append(char* line) {
+  if (_lambdaform_lines == NULL) {
+    _lambdaform_lines = new GrowableArrayCHeap<char*, mtClassShared>(150);
+  }
+  _lambdaform_lines->append(line);
+}
+
+void LambdaFormInvokers::regenerate_holder_classes() {
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/lambdaFormInvokers.cpp:82]: If this is no longer a TRAPS function then it should not be using CHECK macro:

```
objArrayHandle list_lines = oopFactory::new_objArray_handle(vmClasses::String_klass(), len, CHECK);
  for (int i = 0; i < len; i++) {
Handle h_line = java_lang_String::create_from_str(_lambdaform_lines->at(i), CHECK);
```

If exceptions are not possible just use THREAD and comment why they are not possible. If they are possible then you need to handle them appropriately.

Comment (yminqi) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:82]: I will rework the dynamic dump so it will be like static dump --- handle exceptions after returned. So will remove the all the checks in this file. Thanks.

**Comment by yminqi** [comment]
The recent version change made DynamicArchive::dump take TRAP, same as static dump. The design of LambdaFormInvokers::regenerate_holder_classes is that if the regeneration failed, it won't affect normal classes archived. So we have tests for wrong tag or line format in classlist to make sure dump archive. I changed to handle all exceptions except for OOM (which will be processed by caller if possible for the extreme case) in the function.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp [hunk 18]
```diff
@@ -110,12 +120,7 @@ void LambdaFormInvokers::regenerate_holder_classes() {
   JavaValue result(T_OBJECT);
   JavaCalls::call_static(&result, cds_klass, method, signrs, list_lines, THREAD);
 
-  if (HAS_PENDING_EXCEPTION) {
-    log_info(cds)("%s: %s", THREAD->pending_exception()->klass()->external_name(),
-                            java_lang_String::as_utf8_string(java_lang_Throwable::message(THREAD->pending_exception())));
-    CLEAR_PENDING_EXCEPTION;
-    return;
-  }
+  HANDLE_IF_HAS_EXCEPTION;
```
Comment (iklam) [inline src/hotspot/share/cds/lambdaFormInvokers.cpp:123]: Instead of using a macro, I think the code can be more readable like this:

```
if (check_exception(THREAD)) {
    return;
}
```

and `check_exception` can contain the logic of `HANDLE_IF_HAS_EXCEPTION`.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp [hunk 19]
```diff
@@ -79,12 +79,22 @@ void LambdaFormInvokers::append(char* line) {
   _lambdaform_lines->append(line);
 }
 
-void LambdaFormInvokers::regenerate_holder_classes() {
+#define HANDLE_IF_HAS_EXCEPTION                                                                         \
+  if (HAS_PENDING_EXCEPTION) {                                                                          \
+    if (!PENDING_EXCEPTION->is_a(vmClasses::OutOfMemoryError_klass())) {                                \
```
Comment (iklam) [inline src/hotspot/share/cds/lambdaFormInvokers.cpp:84]: I would suggest adding a comment and error message to explain why we ignore the error:

```
// We may get an exception if the classlist contains an error (or an outdated entry generated by
// an older JDK).
if (DumpSharedArchive) {
  log_error(cds)("Failed to generate LambdaForm holder classes. Is your classlist out of date?");
} else {
  log_error(cds)("Failed to generate LambdaForm holder classes. Was the base archive generated with an outdated classlist?");
}
```

File: src/hotspot/share/cds/lambdaFormInvokers.cpp [hunk 20]
```diff
@@ -129,21 +134,11 @@ void LambdaFormInvokers::regenerate_holder_classes() {
...
     ClassFileStream st((u1*)buf, len, NULL, ClassFileStream::verify);
     reload_class(class_name, st, THREAD);
-
-    if (HAS_PENDING_EXCEPTION) {
-      log_info(cds)("Exception happened: %s", PENDING_EXCEPTION->klass()->name()->as_C_string());
-      log_info(cds)("Could not create InstanceKlass for class %s", class_name);
-      CLEAR_PENDING_EXCEPTION;
-      return;
-    }
+    HANDLE_IF_HAS_EXCEPTION;
```
Comment (iklam) [inline src/hotspot/share/cds/lambdaFormInvokers.cpp:141]: Is this necessary? I think bad classlists will cause errors in the JavaCalls::call_static call above, but not here.

**Review by iklam — CHANGES_REQUESTED** [review]

File: src/hotspot/share/cds/lambdaFormInvokers.cpp (hunk 18 above)
Comment (yminqi) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:123]: OK, I was think of that, if a function or macro and finally took macro.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp (hunk 19 above)
Comment (yminqi) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:84]: Do you mean DumpSharedSpaces? So the message will be different between static (DumpSharedSpaces) and dynamic, right?

File: src/hotspot/share/cds/lambdaFormInvokers.cpp (hunk 20 above)
Comment (yminqi) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:141]: Here is the original code, and it is necessary, KlassFactory::create_from_stream could throw exception.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp (hunk 19 above)
Comment (iklam) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:84]: Yes, because the errors can only be caused by bad input from the classlist, which is only used during static archive dump. The information collected during dynamic dump are correct, and cannot cause `CDS.generateLambdaFormHolderClasses()` to fail.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp (hunk 20 above)
Comment (iklam) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:141]: What kid of exceptions will be thrown (other than OOM?) Do you mean `CDS.generateLambdaFormHolderClasses()` can generate bad class files?

Comment (yminqi) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:141]: This is after CDS.generateLambdaFormHolderClasses, we get the byte stream of a class --- we need create instanceklass of the class. The exception is from ClassParser ---- whatever the exception during the process.

Comment (iklam) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:141]: The ClassParser will throw an error only if OOM, or if the class is invalid. However, if the class is invalid, it means there's a bug inside generateLambdaFormHolderClasses() -- it's not supposed to generate an invalid class.

So we cannot just print a warning and ignore the error. The user may not notice the error message. 

The archive dumping should fail so the user knows what has happened, and will hopefully report the error to us.

File: src/hotspot/share/cds/lambdaFormInvokers.cpp [hunk 21]
```diff
@@ -138,7 +140,6 @@ void LambdaFormInvokers::regenerate_holder_classes(TRAPS) {
     memcpy(buf, (char*)h_bytes->byte_at_addr(0), len);
     ClassFileStream st((u1*)buf, len, NULL, ClassFileStream::verify);
     reload_class(class_name, st, THREAD);
```
Comment (iklam) [inline src/hotspot/share/cds/lambdaFormInvokers.cpp:142]: `reload_class(class_name, st, THREAD);` should be `reload_class(class_name, st, CHECK);`. If an error happens inside there, it means `CDS.generateLambdaFormHolderClasses()` has generated a bad classfile, or we have hit an OOM. In either case, we should report the error back to the caller.

**Review by iklam — CHANGES_REQUESTED** [review]

File: src/hotspot/share/cds/lambdaFormInvokers.cpp (hunk 21 above)
Comment (yminqi) [reply src/hotspot/share/cds/lambdaFormInvokers.cpp:142]: Thanks, it should be a check here.

File: src/hotspot/share/cds/lambdaFormInvokers.hpp [hunk 22]
```diff
@@ -26,21 +26,28 @@
...
+  static GrowableArrayCHeap<char*, mtClassShared>* _lambdaform_lines;
+  // for storing LF form lines (LF_RESOLVE only) in read only table.
+  static Array<Array<char>*>* _static_archive_invokers;
   static void reload_class(char* name, ClassFileStream& st, TRAPS);
  public:
-
   static void append(char* line);
+  static void append_filtered(char* line);
+  // All exceptions except for OOM will be suppressed in the fucntion, the
+  // non-OOM failure should not affect normal archiving.
```
Comment (iklam) [inline src/hotspot/share/cds/lambdaFormInvokers.hpp:44]: The above comments should be removed, since we do throw exceptions if the generated classfiles are invalid.

**Review by iklam — APPROVED** [review]
LGTM. Just a small nit about comments.

File: src/hotspot/share/cds/lambdaFormInvokers.hpp (hunk 22 above)
Comment (yminqi) [reply src/hotspot/share/cds/lambdaFormInvokers.hpp:44]: Will update, thanks!

**Review by dholmes-ora — COMMENTED** [review]
The TRAPS/CHECK usages seem okay now.

Thanks,
David

**Comment by yminqi** [comment]
@dholmes-ora Thanks for comment again. @calvinccheung @coleenp I need one more OK to push since this is not a trivial fix.

File: src/hotspot/share/cds/dynamicArchive.cpp [hunk 23]
```diff
@@ -344,29 +345,34 @@ void DynamicArchive::dump(const char* archive_name, TRAPS) {
...
-  }
-  ArchiveClassesAtExit = archive_name;
-  if (Arguments::init_shared_archive_paths()) {
-    dump();
-  } else {
-    ArchiveClassesAtExit = nullptr;
-    THROW_MSG(vmSymbols::java_lang_RuntimeException(),
+    ArchiveClassesAtExit = archive_name;
+    if (Arguments::init_shared_archive_paths()) {
+      DynamicArchive::dump(CHECK);
```
Comment (calvinccheung) [inline src/hotspot/share/cds/dynamicArchive.cpp:350]: I think you can drop the `DynamicArchive::` in the above since it is called from the same class.

**Review by calvinccheung — APPROVED** [review]
Latest version looks good. Just one nit.

Thanks,
Calvin

File: src/hotspot/share/cds/dynamicArchive.cpp (hunk 23 above)
Comment (yminqi) [reply src/hotspot/share/cds/dynamicArchive.cpp:350]: > I think you can drop the `DynamicArchive::` in the above since it is called from the same class.

Thanks. Will remove the DynamicArchive:: at push.
